### PR TITLE
Add class groups

### LINKED ISSUE #25: Create class
### User Story

As a TA of many classes in one module, I can create a class within a module, with information such as venue and date/time, so that I can differentiate students based on their class.

### PR DESCRIPTION
**Pull request by chaiwanlin**
closes #25

### CODE REVIEW

File: src/main/java/seedu/edrecord/logic/commands/MakeGroupCommand.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    public static final String MESSAGE_SUCCESS = "Class %1$s created in %2$s!";
+
+    private final Group group;
+    private final Module module;
+
+    /**
+     * @param group to be created.
+     */
+    public MakeGroupCommand(Group group, Module mod) {
+        requireAllNonNull(group);
```
Comment (caipng) [inline src/main/java/seedu/edrecord/logic/commands/MakeGroupCommand.java:39]: Is this supposed to check for `mod` as well?

File: src/main/java/seedu/edrecord/model/Model.java [hunk 2]
```diff
@@ -114,10 +115,16 @@
 
     /**
      * Adds the given module.
-     * {@code mod} must not already exist in EdRecord.
+     * {@code person} must not already exist in the module system.
```
Comment (caipng) [inline src/main/java/seedu/edrecord/model/Model.java:118]: Is this a typo? `mod` is correct right?

File: src/main/java/seedu/edrecord/model/group/Group.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    public static final String MESSAGE_CONSTRAINTS = "Class code cannot have whitespaces.";
+    public static final String MESSAGE_DOES_NOT_EXIST = "Class with that code has yet to be created.";
+    public static final String MESSAGE_DUPLICATE = "Class with that code has already been created.";
+
+    /*
+     * The class code must not have any whitespace characters.
+     */
+    public static final String VALIDATION_REGEX = "[^\\s]+";
+
+    public static final ModuleSystem GROUP_SYSTEM = new ModuleSystem();
```
Comment (caipng) [inline src/main/java/seedu/edrecord/model/group/Group.java:22]: Is this supposed to be of type `GroupSystem`?

File: src/main/java/seedu/edrecord/model/group/GroupSystem.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+     *   among constructors.
+     */
+    {
+        groups = new UniqueGroupList();
+    }
+
+    public GroupSystem() {}
+
+    /**
+     * Creates an GroupSystem using the Groups in the {@code toBeCopied}
```
Comment (caipng) [inline src/main/java/seedu/edrecord/model/group/GroupSystem.java:31]:
```suggestion
     * Creates a GroupSystem using the Groups in the {@code toBeCopied}
```

**Comment by FergusMok** [comment]
There's no checks to create a person to an existing class for now

Sample test case:
1. add n/Betsy Crowe p/1234567 e/betsycrowe@example.com m/CS2103 c/W-14-3 i/telegram @betyoass t/strong
(W-14-3 is not held by the Module, but the Person has the class)
2. add n/Bob p/1234567 e/betsycrowe@example.com m/CS2103 c/W-3 i/telegram @betyoass t/strong
3. edit <index of Bob> c/W-14-3
> Error Message: Class with that code has yet to be created.

If we're rushing, we can shift this to another issue in 1.3

**Comment by chaiwanlin** [comment]
> There's no checks to create a person to an existing class for now
> 
> Sample test case:
> 
> 1. add n/Betsy Crowe p/1234567 e/betsycrowe@example.com m/CS2103 c/W-14-3 i/telegram @betyoass t/strong
>    (W-14-3 is not held by the Module, but the Person has the class)
> 2. add n/Bob p/1234567 e/betsycrowe@example.com m/CS2103 c/W-3 i/telegram @betyoass t/strong
> 3. edit  c/W-14-3
> 
> > Error Message: Class with that code has yet to be created.
> 
> If we're rushing, we can shift this to another issue in 1.3

These test cases are all under Add..Test or Edit..Test already but I didn't have to change them because the TypicalModules and PersonBuilder takes class into consideration already. However I haven't added test cases for making mods and class I'll put those under v1.3 for now since I might not be able to rush them out yet

**Review by marcusc55 — APPROVED** [review]
LGTM!

**Comment by FergusMok** [comment]
@chaiwanlin Sorry I meant non-existing class, I think it's a non-intended bug. I think test cases not super impt for now

File: src/test/java/seedu/edrecord/logic/commands/EditCommandTest.java [hunk 5]
```diff
@@ -47,6 +47,8 @@ public void execute_allFieldsSpecifiedUnfilteredList_success() {
         Model expectedModel = new ModelManager(new EdRecord(model.getEdRecord()),
                 new ModuleSystem(model.getModuleSystem()), new UserPrefs());
         expectedModel.setPerson(model.getFilteredPersonList().get(0), editedPerson);
+        System.out.println(model.getFilteredPersonList().get(0));
+        System.out.println(editedPerson);
```
Comment (caipng) [inline src/test/java/seedu/edrecord/logic/commands/EditCommandTest.java:51]:
```suggestion
```

File: src/test/java/seedu/edrecord/storage/JsonAdaptedPersonTest.java [hunk 6]
```diff
@@ -40,95 +44,113 @@
...
 
     @Test
     public void toModelType_nullModule_throwsIllegalValueException() {
         JsonAdaptedPerson person = new JsonAdaptedPerson(VALID_NAME, VALID_PHONE, VALID_EMAIL, VALID_INFO,
-                null, VALID_TAGS);
+                null, VALID_GROUP, VALID_TAGS);
         String expectedMessage = String.format(MISSING_FIELD_MESSAGE_FORMAT, Module.class.getSimpleName());
         assertThrows(IllegalValueException.class, expectedMessage, person::toModelType);
     }
+    @Test
```
Comment (caipng) [inline src/test/java/seedu/edrecord/storage/JsonAdaptedPersonTest.java:130]: missing newline

**Review by caipng — CHANGES_REQUESTED** [review]
Some minor nits.

**Review by FergusMok — APPROVED** [review]
LGTM

File: src/main/java/seedu/edrecord/model/Model.java (hunk 2 above)
Comment (chaiwanlin) [reply src/main/java/seedu/edrecord/model/Model.java]: Yeap, updated

File: src/main/java/seedu/edrecord/model/group/Group.java (hunk 3 above)
Comment (chaiwanlin) [reply src/main/java/seedu/edrecord/model/group/Group.java]: Nope, not supposed to be there. Updated

File: src/test/java/seedu/edrecord/logic/commands/EditCommandTest.java (hunk 5 above)
Comment (chaiwanlin) [reply src/test/java/seedu/edrecord/logic/commands/EditCommandTest.java]: Yeap, updated

File: src/test/java/seedu/edrecord/storage/JsonAdaptedPersonTest.java (hunk 6 above)
Comment (chaiwanlin) [reply src/test/java/seedu/edrecord/storage/JsonAdaptedPersonTest.java:131]: Yeap, updated

**Review by caipng — APPROVED** [review]
LGTM!

**Pull request merged**